spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/SequenceBeamSearch.scala

```diff
@@ -0,0 +1,423 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.intel.analytics.bigdl.nn
+
+import com.intel.analytics.bigdl.tensor.TensorNumericMath.TensorNumeric
+import com.intel.analytics.bigdl.tensor._
+
+import scala.collection.mutable.ArrayBuffer
+import scala.reflect.ClassTag
+
+class SequenceBeamSearch[T: ClassTag](
+  val symbolsToLogitsFn: (Tensor[T], Tensor[Int], Map[String, Map[String, Tensor[T]]],
+   Map[String, Tensor[T]]) =>
+   (Tensor[T], Map[String, Map[String, Tensor[T]]], Map[String, Tensor[T]]),
+  val vocabSize: Int,
+  val batchSize: Int,
+  val beamSize: Int,
+  val alpha: Float,
+  val maxDecodeLength: Int,
+  val eosID: Float)(implicit ev: TensorNumeric[T]) {
+
+  def expandDim(tensor: Tensor[T], axis: Int): Tensor[T] = {
+    val shape = tensor.size()
+    val newShape = shape.toBuffer
+    newShape.insert(axis, 1)
+    tensor.reshape(newShape.toArray)
+  }
+
+  // Tiles a given tensor by beam_size.
+  def extendBeamSize(t: Tensor[T], beamSize: Int): Tensor[T] = {
+    val tensor = expandDim(t, 1)
+    val tileDim = new Array[Int](tensor.dim()).map(a => a + 1)
+    tileDim(1) = beamSize
+    tensor.repeatTensor(tileDim)
+  }
+
+  def lengthNormalization(alpha: Float, length: Int): T = {
+    ev.pow(ev.fromType[Double](5.0 + length / 6.0), ev.fromType[Float](alpha))
+  }
+
+  def boolToFloat(b: Boolean): T = {
+    if (b) ev.fromType[Float](1.0f)
+    else ev.fromType[Float](0.0f)
+  }
+
+  def floatToBool(f: T): Boolean = {
+    if (f == 1.0) true
+    else false
+  }
+
+  def reduceAny(tensor: Tensor[Boolean]): Tensor[T] = {
+    val tSize = tensor.size()
+    val outputs = Tensor[T](tSize(0))
+    for (i <- 1 to tSize(0)) {
+      var valueAny = false
+      for (j <- 1 to tSize(1)) {
+        valueAny = valueAny || tensor.valueAt(i, j)
+      }
+      outputs.setValue(i, boolToFloat(valueAny))
+    }
+    outputs
+  }
+
+  def reduceAll(tensor1: Tensor[T], tensor2: Tensor[T]): Boolean = {
+    val sizeT = tensor1.size()
+    var outputs = true
+    for (i <- 1 to sizeT(0)) {
+      outputs &&= ev.isGreater(tensor1.valueAt(i), tensor2.valueAt(i))
+    }
+    outputs
+  }
+
+  def continueSearch(state: Map[String, Any]): Boolean = {
+    val inf = 1.0 * 1e7 * (-1)
+    val i = state("CUR_INDEX").asInstanceOf[Tensor[Int]].clone()
+    val alive_log_probs = state("ALIVE_LOG_PROBS").asInstanceOf[Tensor[T]].clone()
+    val finished_scores = state("FINISHED_SCORES").asInstanceOf[Tensor[T]].clone()
+    val finished_flags = state("FINISHED_FLAGS").asInstanceOf[Tensor[Boolean]].clone()
+    var notAtMaxDecodeLength = true
+    if (i(Array(1)) < maxDecodeLength) {
+      notAtMaxDecodeLength = true
+    } else {
+      notAtMaxDecodeLength = false
+    }
+    val maxLengthNorm = lengthNormalization(alpha, maxDecodeLength)
+    val bestAliveScores = alive_log_probs.select(2, 1).clone() / maxLengthNorm
+    val newFinishedFlags = Tensor[T](finished_flags.size())
```

Comment:
Not create new tensor for each time. You can create a global tensor and just update its values.